### DISCUSSION TITLE
Future data inserted into default partition and maintenance failed.

### DISCUSSION THREAD

**Discussion by pdev91** [pgpartman/pg_partman#619] [Q&A] [answered]
We're running pg_partman on our Zabbix installation with the history tables partitioned daily and the trends tables partitioned monthly. For some reason, a host tried to insert data from 1/12/2029 on 1/12/2024. We don't have that many partitions created so it went into the default partitions for some of these tables. The background worker ran and it failed which seems to have broke partitioning. I'm new to the partitioning/pg_partman world so not totally confident in the ins/outs but I assume this failed because it didn't have the trends_p2029_01 partition already created. We had the next 4 days (premake is 4 for the dailies) created (13-16th) and data was inserted as normal but the Jan 17th partition wasn't created on the 13th and so on which led to millions and even billions of records to be left in the default partition.

I've since removed the 2029 data and started moving the data with partition_data_proc() but this takes a lot of time. Looking for prevention measures.

My questions:

1. Is this normal behavior where if a partition doesn't exist it goes into the default and pg_partman can't pull it out?
2. Is there a way to prevent this from happening besides creating a lot of partitions?
3. How/when are new partitions created?
4. Is the default partition necessary? Wondering if we'd prefer that inserts fail for partitions not existing due to bad data as opposed to trying to move the data from the default.

Error:
```
2024-01-12 20:43:31.503 CST [3600180] ERROR:  updated partition constraint for default partition "trends_default" would be violated by some row
        CONTEXT: SQL statement "ALTER TABLE public.trends ATTACH PARTITION public.trends_p2029_01 FOR VALUES FROM ('1861941600') TO ('1864620000')"
        PL/pgSQL function create_partition_time(text,timestamp with time zone[],boolean,text) line 259 at EXECUTE
        PL/pgSQL function partman.run_maintenance(text,boolean,boolean) line 284 at assignment
        SQL statement "SELECT "partman".run_maintenance(p_analyze := false, p_jobmon := true)"
        DETAIL:
        HINT:
        CONTEXT: PL/pgSQL function create_partition_time(text,timestamp with time zone[],boolean,text) line 511 at RAISE
        PL/pgSQL function partman.run_maintenance(text,boolean,boolean) line 284 at assignment
        SQL statement "SELECT "partman".run_maintenance(p_analyze := false, p_jobmon := true)"
        DETAIL:
        HINT:
2024-01-12 20:43:31.503 CST [3600180] CONTEXT:  PL/pgSQL function partman.run_maintenance(text,boolean,boolean) line 413 at RAISE
        SQL statement "SELECT "partman".run_maintenance(p_analyze := false, p_jobmon := true)"
2024-01-12 20:43:31.506 CST [1074141] LOG:  background worker "????" (PID 3600180) exited with exit code 1
```

Table (post fix):
```
storezabbixdb=# \d+ trends
                                                Partitioned table "public.trends"
  Column   |       Type       | Collation | Nullable |        Default        | Storage | Compression | Stats target | Description
-----------+------------------+-----------+----------+-----------------------+---------+-------------+--------------+-------------
 itemid    | bigint           |           | not null |                       | plain   |             |              |
 clock     | integer          |           | not null | 0                     | plain   |             |              |
 num       | integer          |           | not null | 0                     | plain   |             |              |
 value_min | double precision |           | not null | '0'::double precision | plain   |             |              |
 value_avg | double precision |           | not null | '0'::double precision | plain   |             |              |
 value_max | double precision |           | not null | '0'::double precision | plain   |             |              |
Partition key: RANGE (clock)
Indexes:
    "trends_pkey" PRIMARY KEY, btree (itemid, clock)
Partitions: trends_p2023_09 FOR VALUES FROM (1693544400) TO (1696136400),
            trends_p2023_10 FOR VALUES FROM (1696136400) TO (1698814800),
            trends_p2023_11 FOR VALUES FROM (1698814800) TO (1701410400),
            trends_p2023_12 FOR VALUES FROM (1701410400) TO (1704088800),
            trends_p2024_01 FOR VALUES FROM (1704088800) TO (1706767200),
            trends_p2024_02 FOR VALUES FROM (1706767200) TO (1709272800),
            trends_p2024_03 FOR VALUES FROM (1709272800) TO (1711947600),
            trends_p2024_04 FOR VALUES FROM (1711947600) TO (1714539600),
            trends_p2024_05 FOR VALUES FROM (1714539600) TO (1717218000),
            trends_p2024_06 FOR VALUES FROM (1717218000) TO (1719810000),
            trends_p2024_07 FOR VALUES FROM (1719810000) TO (1722488400),
            trends_p2024_08 FOR VALUES FROM (1722488400) TO (1725166800),
            trends_p2024_09 FOR VALUES FROM (1725166800) TO (1727758800),
            trends_p2024_10 FOR VALUES FROM (1727758800) TO (1730437200),
            trends_default DEFAULT


```
part_config for public.trends:
```
storezabbixdb=# select * from partman.part_config where parent_table = 'public.trends';
-[ RECORD 1 ]--------------+-------------------------------
parent_table               | public.trends
control                    | clock
partition_type             | native
partition_interval         | 1 mon
constraint_cols            |
premake                    | 4
optimize_trigger           | 4
optimize_constraint        | 30
epoch                      | seconds
inherit_fk                 | t
retention                  | 12 month
retention_schema           |
retention_keep_table       | f
retention_keep_index       | t
infinite_time_partitions   | f
datetime_string            | YYYY_MM
automatic_maintenance      | on
jobmon                     | t
sub_partition_set_full     | f
undo_in_progress           | f
trigger_exception_handling | f
upsert                     |
trigger_return_null        | t
template_table             | partman.template_public_trends
publications               |
inherit_privileges         | f
constraint_valid           | t
subscription_refresh       |
drop_cascade_fk            | f
ignore_default_data        | f
```

**Comment by keithf4** [comment, marked as answer]
> 1. Is this normal behavior where if a partition doesn't exist it goes into the default and pg_partman can't pull it out?

Yes, this is the normal behavior when a default partition exists

> 2. Is there a way to prevent this from happening besides creating a lot of partitions?

If you want to have a default partition to avoid losing data, then any data outside the window of available children will get that data inserted to the default. There is no way to prevent it if the available child does not exist.

> 3. How/when are new partitions created?

New partitions are made when the `run_maintenance()` or `run_maintenance_proc()` function is called via some scheduler app (cron, etc). Or the background worker has been set up to run them for you. Check if `shared_preload_libraries` has `pg_partman_bgw` defined in it and if there are any other `pg_partman_bgw.*` parameters defined in your postgresql.conf

> 4. Is the default partition necessary? Wondering if we'd prefer that inserts fail for partitions not existing due to bad data as opposed to trying to move the data from the default.

The default partition is the standard behavior for pg_partman since I wanted its standard behavior to not be to lose data. Whether it is necessary or not is entirely up to you and if you'd rather PostgreSQL throw an error when data outside of the available children is inserted, you are free to drop it. partman will only create the default table when initially setting up the partition set and in versions 5.x and later there is now an option whether to create the default as a parameter to the `create_parent()` function.

Hopefully that answered all you're questions! Please let me know if you need further assistance.

**Reply by pdev91** [reply]
I appreciate your responses. This definitely helps me out greatly!

**Comment by keithf4** [comment]
Also for reference, versions 4.x and earlier did take into account data in the default partition when creating future partitions. This is why you were missing the partitions you needed and it was trying to make partitions in 2029. For version 5.x and later, data in the default partition is ignored for normal child table creation and if a new child table tries to be made that would match data in the default, that is when an error will be thrown. So hopefully if you're able to upgrade to 5.0, data going into the default will be less of a problem if you'd like to keep the default partition around.

**Comment by pdev91** [comment]
Hello, we've upgraded to 5.4.3 and it seems data in the default partition was not ignored for normal child table creation. For some reason, a host rebooted and the time was 3 months in the future so when Zabbix tried to insert it it went into the default. When the maintenance ran, it errored and stopped making the normal daily June partitions. 

Am I supposed to update a setting to enable this "ignore data in the default until that day arrives" behavior?

<img width="1992" height="199" alt="image" src="https://github.com/user-attachments/assets/562e4f73-98f4-41d8-a69a-e402b147b75d" />


`2026-06-06 00:39:36.091 CDT [3260429] @   ERROR:  updated partition constraint for default partition "history_default" would be violated by some row
        CONTEXT: SQL statement "ALTER TABLE public.history ATTACH PARTITION public.history_p2026_09_07 FOR VALUES FROM ('1788757200') TO ('1788843600')"
        PL/pgSQL function create_partition_time(text,timestamp with time zone[],text) line 220 at EXECUTE
        PL/pgSQL function run_maintenance(text,boolean,boolean) line 343 at assignment
        SQL statement "SELECT partman.run_maintenance('public.history', p_jobmon := 'f', p_analyze := 'f')"
        PL/pgSQL function partman.run_maintenance_proc(integer,boolean,boolean) line 42 at EXECUTE
        SQL statement "CALL partman.run_maintenance_proc(p_wait => 0, p_analyze => false, p_jobmon => false)"
        DETAIL:
        HINT:
        CONTEXT: PL/pgSQL function create_partition_time(text,timestamp with time zone[],text) line 401 at RAISE
        PL/pgSQL function run_maintenance(text,boolean,boolean) line 343 at assignment
        SQL statement "SELECT partman.run_maintenance('public.history', p_jobmon := 'f', p_analyze := 'f')"
        PL/pgSQL function partman.run_maintenance_proc(integer,boolean,boolean) line 42 at EXECUTE
        SQL statement "CALL partman.run_maintenance_proc(p_wait => 0, p_analyze => false, p_jobmon => false)"
        DETAIL:
        HINT:
2026-06-06 00:39:36.091 CDT [3260429] @   CONTEXT:  PL/pgSQL function run_maintenance(text,boolean,boolean) line 492 at RAISE
        SQL statement "SELECT partman.run_maintenance('public.history', p_jobmon := 'f', p_analyze := 'f')"
        PL/pgSQL function partman.run_maintenance_proc(integer,boolean,boolean) line 42 at EXECUTE
        SQL statement "CALL partman.run_maintenance_proc(p_wait => 0, p_analyze => false, p_jobmon => false)"
2026-06-06 00:39:36.107 CDT [2318356] @   LOG:  background worker "pg_partman background worker" (PID 3260429) exited with exit code 1
`

**Comment by keithf4** [comment]
Can you share the output for `\d+ public.history` as well as the full output of `select * from partman.part_config where parent_table = 'public.history';`?

**Reply by pdev91** [reply]
```
zabbixdb=# select * from partman.part_config where parent_table = 'public.history'\gx
-[ RECORD 1 ]------------------+--------------------------------
parent_table                   | public.history
control                        | clock
partition_interval             | 1 day
partition_type                 | range
premake                        | 10
automatic_maintenance          | on
template_table                 | partman.template_public_history
retention                      | 30 day
retention_schema               |
retention_keep_index           | t
retention_keep_table           | f
epoch                          | seconds
constraint_cols                |
optimize_constraint            | 30
infinite_time_partitions       | f
datetime_string                | YYYY_MM_DD
jobmon                         | t
sub_partition_set_full         | f
undo_in_progress               | f
inherit_privileges             | f
constraint_valid               | t
ignore_default_data            | f
date_trunc_interval            |
maintenance_order              |
retention_keep_publication     | f
maintenance_last_run           | 2026-06-05 11:39:13.421201-05
time_encoder                   |
time_decoder                   |
async_partitioning_in_progress |
```

```
zabbixdb=# \d+ public.history
                                              Partitioned table "public.history"
 Column |       Type       | Collation | Nullable |        Default        | Storage | Compression | Stats target | Description
--------+------------------+-----------+----------+-----------------------+---------+-------------+--------------+-------------
 itemid | bigint           |           | not null |                       | plain   |             |              |
 clock  | integer          |           | not null | 0                     | plain   |             |              |
 value  | double precision |           | not null | '0'::double precision | plain   |             |              |
 ns     | integer          |           | not null | 0                     | plain   |             |              |
Partition key: RANGE (clock)
Indexes:
    "history_1" btree (itemid, clock)
Partitions: history_p2026_05_06 FOR VALUES FROM (1778043600) TO (1778130000),
            history_p2026_05_07 FOR VALUES FROM (1778130000) TO (1778216400),
            history_p2026_05_08 FOR VALUES FROM (1778216400) TO (1778302800),
            history_p2026_05_09 FOR VALUES FROM (1778302800) TO (1778389200),
            history_p2026_05_10 FOR VALUES FROM (1778389200) TO (1778475600),
            history_p2026_05_11 FOR VALUES FROM (1778475600) TO (1778562000),
            history_p2026_05_12 FOR VALUES FROM (1778562000) TO (1778648400),
            history_p2026_05_13 FOR VALUES FROM (1778648400) TO (1778734800),
            history_p2026_05_14 FOR VALUES FROM (1778734800) TO (1778821200),
            history_p2026_05_15 FOR VALUES FROM (1778821200) TO (1778907600),
            history_p2026_05_16 FOR VALUES FROM (1778907600) TO (1778994000),
            history_p2026_05_17 FOR VALUES FROM (1778994000) TO (1779080400),
            history_p2026_05_18 FOR VALUES FROM (1779080400) TO (1779166800),
            history_p2026_05_19 FOR VALUES FROM (1779166800) TO (1779253200),
            history_p2026_05_20 FOR VALUES FROM (1779253200) TO (1779339600),
            history_p2026_05_21 FOR VALUES FROM (1779339600) TO (1779426000),
            history_p2026_05_22 FOR VALUES FROM (1779426000) TO (1779512400),
            history_p2026_05_23 FOR VALUES FROM (1779512400) TO (1779598800),
            history_p2026_05_24 FOR VALUES FROM (1779598800) TO (1779685200),
            history_p2026_05_25 FOR VALUES FROM (1779685200) TO (1779771600),
            history_p2026_05_26 FOR VALUES FROM (1779771600) TO (1779858000),
            history_p2026_05_27 FOR VALUES FROM (1779858000) TO (1779944400),
            history_p2026_05_28 FOR VALUES FROM (1779944400) TO (1780030800),
            history_p2026_05_29 FOR VALUES FROM (1780030800) TO (1780117200),
            history_p2026_05_30 FOR VALUES FROM (1780117200) TO (1780203600),
            history_p2026_05_31 FOR VALUES FROM (1780203600) TO (1780290000),
            history_p2026_06_01 FOR VALUES FROM (1780290000) TO (1780376400),
            history_p2026_06_02 FOR VALUES FROM (1780376400) TO (1780462800),
            history_p2026_06_03 FOR VALUES FROM (1780462800) TO (1780549200),
            history_p2026_06_04 FOR VALUES FROM (1780549200) TO (1780635600),
            history_p2026_06_05 FOR VALUES FROM (1780635600) TO (1780722000),
            history_p2026_06_06 FOR VALUES FROM (1780722000) TO (1780808400),
            history_p2026_06_07 FOR VALUES FROM (1780808400) TO (1780894800),
            history_p2026_06_08 FOR VALUES FROM (1780894800) TO (1780981200),
            history_p2026_06_09 FOR VALUES FROM (1780981200) TO (1781067600),
            history_p2026_06_10 FOR VALUES FROM (1781067600) TO (1781154000),
            history_p2026_06_11 FOR VALUES FROM (1781154000) TO (1781240400),
            history_p2026_06_12 FOR VALUES FROM (1781240400) TO (1781326800),
            history_p2026_06_13 FOR VALUES FROM (1781326800) TO (1781413200),
            history_p2026_06_14 FOR VALUES FROM (1781413200) TO (1781499600),
            history_p2026_06_15 FOR VALUES FROM (1781499600) TO (1781586000),
            history_default DEFAULT
```

**Comment by keithf4** [comment]
You have `ignore_default_data` set to `false` for this partition set in `part_config`. Try setting that to `true` and run maintenance again.

However, I would also evaluate the data in your default table and see if it's valid or not. Eventually that data will be considered "now" and start throwing this same error when September roles around. If it's invalid data, delete it from the default. If it's valid, run the partition and gap fill functions to partition it out and fill in the gaps.

**Reply by pdev91** [reply]
I appreciate the help. I'll flip this to `true` then. The data that did end up in the default was only for a minute or so. Once the host figured itself out, it went to the correct date/time.

**Reply by pdev91** [reply]
I flipped to `true` and ran the maintenance successfully. Thanks again!
